include/nbl/video/IGPUCommandPool.h

```diff
@@ -334,103 +234,119 @@ class NBL_API IGPUCommandPool : public core::IReferenceCounted, public IBackendO
+                }
+            }
 
-                    prevCmd->~ICommand();
+            return cmd;
+        }
 
-                    // No need to deallocate prevCmd because:
-                    // - it has been allocated from the LinearAddressAllocator where deallocate is a No-OP
-                    // - the memory will get reclaimed in `~LinearAddressAllocator` when whole segment is destroyed
+        // Nullifying the head of the passed segment list is NOT the responsibility of deleteList.
+        inline void deleteList(CCommandSegment* head)
+        {
+            if (!head)
+                return;
 
-                    if (itr.segment == prevSegment)
-                        continue;
+            if (head == m_head)
+                m_head = head->getNextHead();
 
-                    // Segment has changed, we need to free the previous one.
-                    freeSegment(prevSegment);
-                }
+            auto oneBefore = head->getPrevHead();
+            auto oneAfter = head->getNextHead();
+            CCommandSegment::linkHeads(oneBefore, oneAfter);
 
-                // The last segment is always freed outside the loop.
-                assert(tail == prevSegment);
-                freeSegment(tail);
+            if (!oneBefore && !oneAfter)
+                m_head = nullptr;
 
-                headItr.cmd = nullptr;
-                headItr.segment = nullptr;
-                tail = nullptr;
-            }
-            else
+            for (auto& segment = head; segment;)
             {
-                assert(!headItr.segment && !headItr.cmd);
+                auto nextSegment = segment->getNext();
+                segment->~CCommandSegment();
+                m_pool.deallocate(segment, COMMAND_SEGMENT_SIZE);
+                segment = nextSegment;
             }
         }
 
-        void clear(CCommandSegment*& head, CCommandSegment*& tail)
+        inline void clear()
         {
-            // The way deleteList is implemented, we also need to keep the tail pointer of each child command segment list.
-            // See my comment for one possible way to solve this:
-            // https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#issuecomment-1345979974
+            for (auto* currHead = m_head; currHead;)
+            {
+                auto* nextHead = currHead->getNextHead();
+                // We don't (and also can't) nullify the tail here because when the command buffer detects that its parent pool has been resetted
+                // it nullifies both head and tail itself.
+                deleteList(currHead);
+                currHead = nextHead;
+            }
 
-            // for (auto& segment = head; segment; segment = segment->getNextHead())
-            //     deleteList({}, );
+            m_head = nullptr;
         }
 
     private:
+        inline bool appendToList(SCommandSegmentList& list)
+        {
+            auto segment = m_pool.emplace<CCommandSegment>();
```

Comment:
and who's gonna set `list.tail`'s `next` to segment?

This is why I said to do that in `CCommandSegment` constructor, like so
```cpp
        CCommandSegment(CCommandSegment* prev)
        {
            static_assert(alignof(ICommand) == COMMAND_SEGMENT_ALIGNMENT);
            m_header.commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, alignof(ICommand), STORAGE_SIZE);
            m_header.next = nullptr;
            wipeNextCommandSize();
            if (prev) prev->m_header.next = this;
        }
```
and
```cpp
auto segment = m_pool.emplace<CCommandSegment>(list.tail);
```